### PR TITLE
Add validators table

### PR DESCRIPTION
**Pull request by eevui**
Following schema and discussion at https://github.com/MetricsDAO/harmony_dbt/issues/73

### Testing

```txt
root@03e8142ce188:/harmony# dbt run --select validators
Running with dbt=0.21.1
Found 35 models, 284 tests, 0 snapshots, 0 analyses, 178 macros, 0 operations, 8 seed files, 6 sources, 0 exposures

14:10:33 | Concurrency: 4 threads (target='dev')
14:10:33 |
14:10:33 | 1 of 1 START incremental model DEV.validators........................ [RUN]
14:10:53 | 1 of 1 OK created incremental model DEV.validators................... [SUCCESS 1 in 20.30s]
14:10:53 |
14:10:53 | Finished running 1 incremental model in 34.62s.

Completed successfully

Done. PASS=1 WARN=0 ERROR=0 SKIP=0 TOTAL=1
```

image

### Future

Currently, makes direct reference to `harmony.dev.stg_delegators`. Perhaps an issue can be created to make a change when `stg_delegators` is commited to production.

### CODE REVIEW

**Comment by eevui** [comment]
Waiting on #86 , which also resolves making a direct reference to a table.

File: models/core/validators.sql [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+	config(
+		materialized='incremental',
+		unique_key='log_id',
+		tags=['core'],
+		cluster_by=['day_date']
+	)
+}}
+
+with delegators_incremental as (
+	select *
```
Comment (antonyip) [inline models/core/validators.sql]:
```suggestion
	select
		*
```

File: models/core/validators.sql [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+		validator_address,
+		js_onetohex(validator_address) as validator_hex_address,
+		validator_identity,
+		active_status,
+		booted_status,
+		total_delegation as total_one_delegated,
+		ifnull(total_one_rewarded, 0) as total_one_rewarded,
+		ifnull(total_one_undelegated, 0) as total_one_undelegated
+	from delegators_incremental
+	join totals using (ingest_timestamp, validator_address)
```
Comment (antonyip) [inline models/core/validators.sql]: Let's not use using - https://community.snowflake.com/s/question/0D50Z00008WRZBBSA5/bug-with-join-using-

File: models/core/validators.sql [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    group by ingest_timestamp, validator_address
+),
+
+totals as (
+    select
+        ingest_timestamp,
+        validator_address,
+        rewards.amount as total_one_rewarded,
+        undelegations.amount as total_one_undelegated
+    from rewards join undelegations using (ingest_timestamp, validator_address)
```
Comment (antonyip) [inline models/core/validators.sql]: Let's not use using - https://community.snowflake.com/s/question/0D50Z00008WRZBBSA5/bug-with-join-using-

File: models/core/validators.yml [hunk 4]
```diff
@@ -0,0 +1,59 @@
+version: 2
+
+models:
+  - name: validators
+    description: |-
+      This table records various daily statistics of Harmony validators.
+
+    columns:
+      - name: log_id
```
Comment (antonyip) [inline models/core/validators.yml]: lets call this `u_key` -- its not really a log

File: models/core/validators.sql [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+),
+
+undelegations as (
+    select
+        ingest_timestamp,
+        validator_address,
+        sum(ifnull(flat_undelegations.value:amount, 0)) as amount
+    from latest_day_flat_delegations
+	inner join lateral flatten(input => value:undelegations, outer => true) flat_undelegations
+    group by ingest_timestamp, validator_address
```
Comment (antonyip) [inline models/core/validators.sql]:
```suggestion
    group by 1,2
```

File: models/core/validators.sql [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+	inner join lateral flatten(input => delegations)
+),
+
+rewards as (
+    select
+        ingest_timestamp,
+        validator_address,
+        sum(value:reward) as amount
+    from latest_day_flat_delegations
+    group by ingest_timestamp, validator_address
```
Comment (antonyip) [inline models/core/validators.sql]:
```suggestion
    group by 1,2
```

File: models/core/validators.sql [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+	from {{ ref('stg_delegators') }}
+	where {{ incremental_load_filter('ingest_timestamp') }}
+),
+
+latest_day_timestamps as (
+    select
+        max(ingest_timestamp) as ingest_timestamp,
+        validator_address
+    from delegators_incremental
+    group by day_date, validator_address
```
Comment (antonyip) [inline models/core/validators.sql]: @chuxinh hey, is there a way to do group by 1,2 if you don't actually select the columns?

**Comment by antonyip** [comment]
Looks good overall! I only have a small issue with `using`.

File: models/core/validators.sql (hunk 7 above)
Comment (chuxinh) [reply models/core/validators.sql]: nope i don't think `day_date` is needed here
have you tried running dbt on this? it will tell u it if goes through

Comment (antonyip) [reply models/core/validators.sql]: day_date is needed here because we're trying to take the latest daily snapshot..

Comment (chuxinh) [reply models/core/validators.sql]: actually i was wrong, it seems to work? 
but if you need the `day_date`, why not just add that in the select too?

Comment (antonyip) [reply models/core/validators.sql]: @eevui :D.. add day_date, change to group by 1,2

Comment (eevui) [reply models/core/validators.sql]: Should I add `day_date` to the select? I only wanted to treat this CTE as a "primary key" of sorts and `day_date` would be repeated unnecessarily.

Comment (eevui) [reply models/core/validators.sql]: @antonyip pushed a couple commits addressing this

**Comment by eevui** [comment]
To add, just discovered that the incremental filter doesn't work. Is this expected? I get an error saying something along the lines of "aggregate query should be used with `having`" or something like that.

**Comment by antonyip** [comment]
Made a PR into your branch - https://github.com/eevui/harmony_dbt/pull/1

**Review by antonyip — APPROVED** [review]
Approve!

**Pull request merged**